### ISSUE TITLE
Very odd return behavior with a bool return type

### ISSUE THREAD

**Issue by ggryschuk** [Koromix/koffi#159]
First off my utmost apologies for such a long post, especially as my first here and especially as I very much appreciate you providing this library...with that said.

I am converting a rather extensive code base from ffi-lxe (ffi-napi) to use your library as I recently discovered build issues with ffi-napi with node >18.

Anyway, I'm getting my head around how your library marshall's parameters and the like, things were going fine & then I ran in to the oddest of behaviors when attempting to return a 'false' value from a C function. It's so odd I can't even imagine what could be causing this. I'll try to describe it the best I can but the whole thing has me stumped.

My basic usage is that I'm wrapping a 3rd party (open source) C++ library in my own function calls that I compile as C++ but are exposed using 'extern C' as C function calls for use in javascript. Things were going ok in the migration until I tested the following function defined as:

`bool DCM_DCMObject_readFromFile(const char* filename, void** imageDataSet, struct T_Condition* tCond);`

The struct T_Condition is one of my own defined as:

```
struct T_Condition {
	unsigned short module;
	unsigned short result;
	unsigned int status;
	const char* reason;
};

```
In javascript I've set up a test as follows:

```
const koffi = require('koffi');
const path = require('path');

const T_Condition = koffi.struct('T_Condition', {
	module: 'unsigned short',
	result: 'unsigned short',
	status: 'unsigned int',
	reason: 'string'
});

const T_DataSet = koffi.opaque('T_DataSet');
let libPath = __dirname + '/..'
libPath = path.normalize(libPath);
libPath = libPath + '/build/Release/dcmtkjs.so'
const lib = koffi.load(libPath);
const readFromFile = lib.func('DCM_DCMObject_readFromFile','bool',['string','_Out_ T_DataSet**','_Out_ T_Condition*']);
let dataSet = [null];
let tcond = {}
console.log("successfully read file: %s",readFromFile('testfile',dataSet,tcond));
console.log("returned condition: %o",tcond);

```
Now here's where things go sideways. Note that the file 'testfile' above does NOT exist. I was testing for failure. The print out from the console.log line with 'successfully read file: %s' SHOULD have been 'successfully read file: false' but instead I get.

dcmtkjs: DCM_DICOM_readFromFile ERROR: 0001:0012 No such file or directory
successfully read file: true
returned condition: {
  module: 1,
  result: 18,
  status: 1,
  reason: 'No such file or directory'
}

Note how the output from the 'tcond' variable correctly identifies the reason for failure as 'No such file or directory'. Further note that the line that starts with 'dcmtkjs' is simply from a 'printf' statement in the C++/C code.

Since I control the C++/C level source code for the DCM_DCMObject_readFromFile function I traced the 'offending' behavior and did many, many tests and none of this makes much sense. For example I can just do a 'return false' immediately in the C/C++ function & the result will print out as false from the javascript code. I have a test of the filename variable to confirm its not null so if I just pass a 'null' and in javascript it again correctly prints out 'false' on the 'successfully read file' line.

The code block where the 'offending' behavior is originating in the C++ function looks like this (a bit of pseudo-code here):

```
if(cond.bad()) {
     printf(<print some stuff out>);
     return false;                 <---- HERE'S where the return occurs from the C++/C function when the 'file is not found'
}

```
The 'cond' variable above is a C++ object exposed from the 3rd party library & I only use it internally to my function and it is never exposed outside my function. Long story short I can do a 'return false' anywhere outside that C++ code block & in javascript it will properly receive 'false'.

And it gets more interesting from here. If I redefine my C++/C function to return 'int' instead of 'bool' e.g. as:

`int DCM_DCMObject_readFromFile(const char* filename, void** imageDataSet, struct T_Condition* tCond);`

And then update the C++/C code to return 0 (for false) or 1 (for true) BUT I leave the definition of the function the same in javascript (so the lib.func line above doesn't change) I get the CORRECT result. In other words if I define the return value of my C++/C function as 'int' but tell Koffi it's a 'bool' then koffi will return the result as a bool and get it correct (e.g. return 'false')! Of course if I change my definition in javascript to say the function returns an 'int' then koffi will return a '0' (instead of false).

I have many other C++/C functions that I define as returning 'bool' (I haven't tested them all yet) and it wouldn't be hard for me to change their signatures to using 'int' instead of 'bool' but I wouldn't think I would have to. I would normally download a packages code & walk myself through it to try to help figure out what the problem might be here and I tried that with Koffi and for the life of me I can't make heads or tails of the C++/C level code of Koffi.

Long story short, while I seem to have a work around that will make my code work with Koffi I thought I should at least raise this issue here as it's just so very odd. I've experienced things like a C/C++ compiler 'optimizing' code such that I didn't get the behavior I had expected but this doesn't seem like that kind of thing. Plus this code has worked 'as is' with ffi-lxe (ffi-napi) and I use the same node-gyp build methodology.

Anyway, thanks for reading this far.

**Comment by Koromix** [comment]
This was a bug in Koffi regarding the x86_64 ABI, which allows bit 8 to 31 (or 63 for 64-bit registers) to be left as-is for bool values.

Most times, when a bool is returned, the `%eax` register contains 0 or 1. But per-ABI, only `%al` is expected to be exactly 0 or 1. And an optimizing compiler is perfectly free to let some garbage in bits 8 to 31 of `%eax`.

I've fixed Koffi to respect this behavior and only look at the least significant bit. Upgrade to **Koffi 2.8.9** (which I've just published) and this should fix your problem.

**Comment by Koromix** [comment]
Great, thanks for the confirmation 😉

**Comment by ggryschuk** [comment]
So I have a question about the use of your package with ref/ref-napi. It's not an 'issue' other then for me getting my head around using those packages with yours so I'm reticent to post it here. Is it ok if I do or is there a better way to send such a question to you?

**Comment by Koromix** [comment]
No, go ahead, shoot :)

**Comment by ggryschuk** [comment]
Thanks. And as it turns out my question doesn't appear to be about using ref with your library so much as using Buffer's as Output parameters.

As noted I'd like to migrate from ffi-napi to koffi but I make extensive use of ref-napi allocated Buffers as pointers-to-pointers to opaque objects (or pointers to opaque objects). In the example you have wiith sqllite3 you show the use of an array to hold the result of a pointer to an opaque object being returned. E.g. as

```
let out = [null];
if (sqlite3_open_v2(':memory:', out, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, null) != 0)
    throw new Error('Failed to open database');
let db = out[0];
sqlite3_close_v2(db);

```
When I did a console.log(out) of such an array I see it as [External <number>] and its possible that has me confused. I followed that example through my own code and was able to retrieve allocated objects, update the object at that allocate memory etc. by passing this array around as a 'pointer-to-pointer'. But for my purposes I want to use Buffers to hold the pointer returned, specifically "rec.alloc'd" Buffers and it appears this is possible but I'm not 100% sure I understand properly.

Specifically I tested multiple things but at the 'lowest level' I tested an example which amounts to "let out = Buffer.alloc(8") then <call C-Function to allocate memory & return a pointer> in this Buffer which worked. That would be awesome if that's the way its supposed to work as it should make it trivially easy for me to replace loading my C-functions using ffi-napi with koffi.

Long story short, I don't want to celebrate too much yet but does it make sense that I can use a Buffer object to act as the storage location for a 'pointer-to-pointer to opaque' instead of a null initialized array? The caveats you have in your description of ['Output Buffers'](https://koffi.dev/output#output-buffers) suggest this is appropriate but I don't want to assume so. In other words the pointer stored to the Buffer will only be used within the C-functions & I don't manipulate the memory at that location in any way in Javascript (e.g. 'outside the native code').

Hopefully this makes some sense.

thanks.

**Comment by ggryschuk** [comment]
Hi. I've basically answered my own own question about using ref-napi Buffers with koffi. My code works 'as is' with koffi as a 'drop-in' replacement for ffi-napi and passing Buffers for opaque pointers and even Structs. But there is at least a 10x slowdown in performance. I've traced at least one reason to the way that koffi must be populating a 'void pointer' embedded in a Struct. I won't call it an 'issue' per se as I don't think koffi is doing anything wrong, e.g. not a 'bug' at any rate but I'll post it as an issue with a longer description in the hopes that perhaps you can provide an enhancement to koffi.decode or perhaps advise how I should be handling things.

Thanks again for your package, I should be able to replace ffi-lxe/ffi-napi if I can figure out a way to address this performance issue.
